2145/application search requirements

ColinBuyck opened the pull request:
# Pull Request Template

## Issue Overview

This PR addresses #issue

- [x] This change addresses the issue in full
- [ ] This change addresses only certain aspects of the issue
- [ ] This change is a dependency for another issue
- [ ] This change has a dependency from another issue

## Description
The corresponding issue is #2145 

This fix hides the "no rows to show" when filtering on the applications for a specific property on the partner site. Instead, a loading overlay is shown when appsLoading is true and the search has a length greater than 2 characters.

The majority of the changes, however, are focused on requiring the user to enter at least 3 characters before executing the filtering. On the backend, there is now a custom validator that throws a 400 message that details this specific the error and backend tests to ensure consistency on this. The UX is primarily driven by the length of the filter field rather than this 400 error. I went this direction since the useApplicationsData hook resulted in appsData and appsError updating asynchronously and with many intermediary undefined states that made for a poor UX. 

Update: The most recent changes have pertained to refactoring the ApplicationsList function. First, I reordered the functions of the file to be sectioned more clearly. Then, I removed the useApplicationsData hook and instead, refactored the hook body into a new useEffect within the ApplicationsList function that doesn't not use SWR. This useEffect allows for fewer renders than the previous implementation. The applications and its related data were turned into states for clearer and more performant code.

With this new structure, the user alert describing the search length requirements is set on the same delay as the searchFilter to ensure that the user is done typing before calling the backend or showing the alert. This also avoids the confusing use of both filterField and delayedFilter as dynamic values in the previous solution. The 400 error is still in place as a secondary protection. 

## How Can This Be Tested/Reviewed?

This can be tested by signing in to the admin account on the partners site, clicking on a property's applications, and then entering strings of different lengths into the filter

## Checklist:

- [x] My code follows the style guidelines of this project
- [ ] I have added QA notes to the issue with applicable URLs
- [x] I have performed a self-review of my own code
- [x] I have reviewed the changes in a desktop view
- [x] I have reviewed the changes in a mobile view
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [x] I have assigned reviewers
- [ ] I have run `yarn generate:client` and/or created a migration if I made backend changes that require them
- [ ] I have exported any new pieces added to ui-components
- [x] My commit message(s) is/are polished, and any breaking changes are indicated in the message and are well-described
- [ ] Commits made across packages purposefully have the same commit message/version change, else are separated into different commits

## Reviewer Notes:

Steps to review a PR:

- Read and understand the issue, and ensure the author has added QA notes
- Review the code itself from a style point of view
- Pull the changes down locally and test that the acceptance criteria is met
- Also review the acceptance criteria on the Netlify deploy preview (noting that these do not yet include any backend changes made in the PR)
- Either explicitly ask a clarifying question, request changes, or approve the PR if there are small remaining changes but the PR is otherwise good to go

## On Merge:

If you have one commit and message, squash. If you need each message to be applied, rebase and merge.

emilyjablonski commented on sites/partners/pages/listings/[id]/applications/index.tsx:
```diff
@@ -258,7 +300,11 @@ const ApplicationsList = () => {
-
+              <div className="mt-2">
+                {!validSearch && (
+                  <AlertBox type="notice">Enter at least 3 characters to search</AlertBox>
```
This string will need to be added to the partners/page_content/local_overrides/general.json file

emilyjablonski reviewed (changes requested):
This is looking so great! I love the documentation and organization.

I'd love to see a frontend validation check in addition to the backend check on the 3 character limit, because as it works right now, I type 2 characters and hit enter, there is a short delay as we wait for the request, and then we can show the error. But in this case, we don't have to spend resources querying the backend at all because we know we don't have enough characters, so we can use the form field to validate that which will also let us immediately show the error message.

I'd also take a look at the mobile view (though the below is from a small desktop, it happens pretty early in the width).

<img width="757" alt="Screen Shot 2022-04-11 at 6 31 31 PM" src="https://user-images.githubusercontent.com/65367387/162856399-04eecf87-fbb2-41da-9622-875f8c29a152.png">

ColinBuyck commented:
Thank you for the review! The validation is currently done by the frontend and I am pretty sure the delays that you were experiencing were from the debounce function on the filter field. Looking back at it, the delay does seem too long though so I reduced it to half of a second. I tested without a delay and didn't like how it flashed on the screen every time a user began typing. Wondering what your thoughts are on this decision? I also added some styling so the button remains closer to the search field rather than centered(seems more likely to catch users attention) and then wraps below similar to the two buttons for smaller screens. Lastly, added some backend testing to ensure consistency on backend length validation as well.

seanmalbert commented on sites/partners/pages/listings/[id]/applications/index.tsx:
```diff
@@ -37,59 +37,119 @@ type ApplicationsListSortOptions = {
   })
+  useEffect(() => {
+    setAppsLoading(true)
+    const queryParams = new URLSearchParams()
+    queryParams.append("listingId", listingId)
+    queryParams.append("page", currentPage.toString())
+    queryParams.append("limit", itemsPerPage.toString())
+    if (delayedFilterValue) {
+      queryParams.append("search", delayedFilterValue)
+    }
+    if (sortOptions.orderBy) {
+      queryParams.append("orderBy", delayedFilterValue)
+      queryParams.append("order", sortOptions.order ?? EnumApplicationsApiExtraModelOrder.ASC)
+    }
```
Please remove.

emilyjablonski commented:
@ColinBuyck I think showing the invalid search error as soon as we can / have the information to is a better UX than waiting for the debounce, when we know already they don't have enough characters. Quick feedback is usually preferable.

I think it'll mean changing the useEffect to look somethin like this:
`if (filterField.length === 0 || filterField.length > 2) {
      setValidSearch(true)
      debounceFilter.current(filterField)
    } else {
      setDelayedFilterValue("")
      setValidSearch(false)
    }`

dominikx96 reviewed (changes requested):
@ColinBuyck this looks great! I've noticed a few things:

1. We need to apply a similar logic for listings: https://github.com/bloom-housing/bloom/issues/2655 so I think it would be better to create a new, reusable hook for search component (to store input value, debounced value and do not repeat logic everywhere or even a new component that includes search and other components from the table header as props).
2. Multiple re-renders is a good point, but `useSWR` gives us local caching - I see that currently it doesn't work and this hook does a new request every time, but it should store cache and remember previous results - maybe it would be better to keep useSWR and just resolve a problem with caching?

seanmalbert commented:
> @ColinBuyck this looks great! I've noticed a few things:
> 
> 1. We need to apply a similar logic for listings: #2655 so I think it would be better to create a new, reusable hook for search component (to store input value, debounced value and do not repeat logic everywhere or even a new component that includes search and other components from the table header as props).
> 2. Multiple re-renders is a good point, but `useSWR` gives us local caching - I see that currently it doesn't work and this hook does a new request every time, but it should store cache and remember previous results - maybe it would be better to keep useSWR and just resolve a problem with caching?

@dominikx96 , I think item one, you requested is out of scope of this issue, which was to not search unless there were three or more characters, and I don't want the scope of this to keep expanding, so I think that work would be better served in #2655.

> * Did you consider using the page wide notification that would appear above the table?
> * Can we generalize/repurpose this component for more than this validation? Would be curious to use this component for any validation on this field not just search length

@slowbot , I think the concern with using the page wide notification is that it can cause the page to move up and down while typing, unless we left it up for the user to close. As for generalizing the component, @ColinBuyck used the `AlertBox`, so we could use that elsewhere in a similar fashion.

emilyjablonski reviewed (approved):
This looks great! I know it was bigger than anticipated - it's a really nice UX improvement.
